### PR TITLE
msg/async/Protocol*: send keep alive if existing wins

### PR DESCRIPTION
**Pull request by xiexingguo**
This is a follow-up fix of https://github.com/ceph/ceph/pull/25754.

It turns out the existing connection would win the connection race
and then be stuck at the __replacing__ stage forever without being
aware of the underlying Pipe is actually broken.

Fix by forcing existing sending keepalive periodically __too__.

Fixes: http://tracker.ceph.com/issues/38493
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by liewegas** [comment]
needs rebase

File: src/msg/async/ProtocolV1.cc [hunk 1]
```diff
@@ -1999,6 +1999,10 @@ CtPtr ProtocolV1::handle_connect_message_2() {
                     << connection->get_state_name(existing->state) << dendl;
       reply.global_seq = exproto->peer_global_seq;
       existing->lock.unlock();
+      // existing connection wins
+      // make sure pipe is still functioning before we can safely
+      // follow through with it.
```
Comment (liewegas) [inline src/msg/async/ProtocolV1.cc]: maybe just
```
// make sure existing connection is still functioning
```

**Comment by xiexingguo** [comment]
@liewegas Comment fixed, branch re-pushed.

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2019-02-28_06:43:32-rados-wip-38493-distro-basic-smithi/

QA looks fine.

File: src/msg/async/ProtocolV2.cc [hunk 2]
```diff
@@ -2464,6 +2464,10 @@ CtPtr ProtocolV2::handle_reconnect(ceph::bufferlist &payload)
     ldout(cct, 1) << __func__
                   << " existing racing replace happened while replacing."
                   << " existing=" << existing << dendl;
+    existing->lock.unlock();
+    // make sure we notice if existing connection is no longer functioning
+    existing->send_keepalive();
+    existing->lock.lock();
```
Comment (liewegas) [inline src/msg/async/ProtocolV2.cc]: I think both of these call sites dropping lock will introduce bugs.. any time the lock is dropped we need to reverify the connection state.

Instead, can we make an locked variant of send_keepalive?

File: src/msg/async/ProtocolV2.cc [hunk 3]
```diff
@@ -2571,6 +2575,10 @@ CtPtr ProtocolV2::handle_existing_connection(AsyncConnectionRef existing) {
...
```
Comment (liewegas) [inline src/msg/async/ProtocolV2.cc]: here too

File: src/msg/async/ProtocolV2.cc (hunk 2 above)
Comment (xiexingguo) [reply src/msg/async/ProtocolV2.cc]: @liewegas I copied these code from https://github.com/ceph/ceph/pull/26668/files#diff-ea445419e9f7f9529b73ce454e5e3c53R2657

But now I am getting confused. Why do we need to drop the lock before calling 
`send_keepalive()`? Can't we simply call `send_keepalive()` under the scope of lock?

Comment (liewegas) [reply src/msg/async/ProtocolV2.cc]: Ah, it looks like it does work fine with the lock held (it only takes write_lock).  But i suspect the other call site is broken in the same way, then (it shouldn't drop the lock!).  Can you fix it as well?

Comment (xiexingguo) [reply src/msg/async/ProtocolV2.cc]: @liewegas Sure. Fixed!

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2019-03-02_04:02:56-rados-wip-38493-distro-basic-smithi/

**Pull request merged**